clang/test/SemaSYCL/device_global.cpp

```diff
@@ -18,21 +19,82 @@ device_global<char> Foo::d;
+  // expected-error@+1{{'device_global' member variable 'f' should be publicly accessible from namespace scope}}
   static device_global<int> f;
 };
+
 device_global<int> Baz::f;
 
 device_global<int[4]> not_array; // OK
 
+// expected-error@+1{{'device_global' array is not allowed}}
+device_global<int> array[4];
+
 device_global<int> same_name; // OK
+
 namespace foo {
 device_global<int> same_name; // OK
 }
-namespace {
-device_global<int> same_name; // OK
+
+struct BBar {
+private:
+  struct BarInsider {
+    // expected-error@+1{{'device_global' member variable 'c' should be publicly accessible from namespace scope}}
+    static device_global<float> c;
+  };
+};
+
+struct ABar {
+  void method() {
+    // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
+    static device_global<float> c;
+  }
+  struct BarInsider {
+    static device_global<float> c;
+    void method() {
+      // expected-error@+1{{'device_global' variable must be a static data member or declared in global or namespace scope}}
+      static device_global<float> c;
+    }
+  };
+};
+
+template <typename T> void fooBar() {
+  // expected-error@+2 {{'device_global' member variable 'c' should be publicly accessible from namespace scope}}
+  // expected-error@+1 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  static device_global<T> c;
+  // expected-error@+2 {{'device_global' member variable 'd' should be publicly accessible from namespace scope}}
+  // expected-error@+1 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  device_global<T> d;
 }
 
+template <typename T> struct TS {
+private:
+  // expected-error@+1 2 {{'device_global' member variable 'a' should be publicly accessible from namespace scope}}
+  static device_global<T> a;
+  // expected-error@+1 2 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  device_global<T> b;
+  // expected-error@+2 {{'device_global' member variable 'c' should be publicly accessible from namespace scope}}
+  // expected-error@+1 2 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  device_global<int> c;
+
+public:
+  static device_global<T> d;
+  // expected-error@+1 2 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  device_global<T> e;
+  // expected-error@+1 2 {{'device_global' variable must be a static data member or declared in global or namespace scope}}
+  device_global<int> f;
+};
+
+// expected-note@+1 {{in instantiation of template class 'TS<int>' requested here}}
+TS<int> AAAA;
+
+//expected-note@+2 {{in instantiation of template class 'TS<char>' requested here}}
+template <typename T> void templFoo () {
+  TS<T> Var;
+}
+
+template<> void templFoo<int>();
```

Comment:
Sorry I didn't catch this earlier but is this template specialization declaration required? I believe @Fznamznon's request was to just instantiate and call function, which you have done in main() below.